apps/indexer/src/api/services/historical-balances/historical-balances.service.ts

```diff
@@ -0,0 +1,272 @@
+import { Address, createPublicClient, http, PublicClient, encodeFunctionData } from "viem";
+import { readContract } from "viem/actions";
+
+import { DaoIdEnum } from "@/lib/enums";
+import { CONTRACT_ADDRESSES } from "@/lib/constants";
+import { getChain } from "@/lib/utils";
+import { env } from "@/env";
+
+export interface HistoricalBalance {
+  address: Address;
+  balance: bigint;
+  blockNumber: number;
+  tokenAddress: Address;
+}
+
+export interface HistoricalBalancesRequest {
+  addresses: Address[];
+  blockNumber: number;
+  daoId: DaoIdEnum;
+}
+
+// Multicall3 contract address (deployed after governance setup)
+const MULTICALL3_ADDRESS = "0xA51c1fc2f0D1a1b8494Ed1FE312d7C3a78Ed91C0" as Address;
```

Comment:
this already exists within [viem's chain config](https://github.com/wevm/viem/blob/main/src/chains/definitions/arbitrum.ts#L20)